security/security_principles/README.md

```diff
@@ -0,0 +1,76 @@
+
+Handling errors securely is a key aspect of secure coding. Exceptions that occur while processing a security control deserve special attention. These exceptions must not allow behavior that the control wouldn’t normally allow. Any time we implement a security mechanism, there are three possible outcomes:
+
+Allow the operation
+Disallow the operation
+Exception
+
+Generally, we design security mechanisms so that failure, or exception, is the same as disallowing the operation.
+
+## Principle of Least Privilege
+
+**A secure application gives the least amount of privilege required.**
+
+This principle encompasses user rights and resource permissions.
+
+For example, when we add clients to Slack, we only give them access to one external client channel.
+
+Another example is a server that only needs access to a network, read access to a database table, and the ability to write a log. So those are the only permissions that the server should be granted.
+
+Keep in mind that it’s often easier to give additional permissions than to take away existing permissions.
+
+## Separation of Duties
+
+**A secure application separates duties, restricting the power of any individual.**
+
+Separation of duties manages conflict of interest by restricting any one individual’s power. It is best implemented by using role-based access control and by carefully defining what each role should be able to do.
+
+We can apply this method to job roles, where a developer develops, a QA tester tests, and a devops engineer deploys. This kind of separation ensures plenty of eyes on the code before it goes live.
+
+Another example of this method is how we write code at Sparkbox. We write modular, componentized code to limit complexity, which can limit what an attacker may be able to communicate with in an attack.
+
+## Security by Obscurity
+
+**A secure application is not reliant upon keeping details hidden.**
+
+Security through obscurity means “don’t try to hide things.” Attackers have all the time in the world to find things we hide from them. Security by obscurity is a weak security control, meaning that systems shouldn’t be reliant upon keeping details hidden.
+
+For example, the security of a web application shouldn’t rely on knowledge of the source code being kept secret. That’s not to say keeping secrets is bad, but it’s not a reliable way to secure a web application on it’s own—be sure to use multiple levels of security on top of simply hiding information.
+
+## Do Not Trust the Client
+
+**A secure application validates information coming from the client.**
+
+This principle acknowledges that anything you send a client (a person or machine that’s consuming the application) can be changed and sent back. For this reason, validation should be performed in workflow or command patterns even if backend code performs validation as well. This validation process prevents unnecessary round trips to the backend and prevents the backend from revealing most tampered data.
```

Comment:
The first statement is definitely accurate... what you send to the client can be tampered with. In fact, data sent to the server might not even come from your client code. (Postman is a good example of this).

For that reason, I think we can clarify this a little:

```suggestion
This principle acknowledges that anything you send a client (a person or machine that’s consuming the application) can be tampered with. Data sent from the client to the server should not be trusted. For this reason, the backend code should validate all data and permissions. 
```

I'm not sure I understand this.
> This validation process prevents unnecessary round trips to the backend and prevents the backend from revealing most tampered data.

Reply:
Yep this adjustment is way more clear 👍